File: 0129/solution.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function sumNumbers(root: TreeNode | null): number {
  const traverse = (node: TreeNode | null, prefix: number): number => {
    if (!node) {
      return 0;
    }

    if (!node.left && !node.right) {
      return prefix * 10 + node.val;
    }

    return (
      traverse(node.left, prefix * 10 + node.val) +
      traverse(node.right, prefix * 10 + node.val)
    );
  };
  return traverse(root, 0);
}

File: 0129/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        res = 0

        def root_to_leaf(node, val):
            curr = node.val + val
            if not node.left and not node.right:
                nonlocal res
                res += curr
            curr *= 10
            if node.left:
                root_to_leaf(node.left, curr)
            if node.right:
                root_to_leaf(node.right, curr)

        root_to_leaf(root, 0)
        return res
